### PR TITLE
[Example] Implement LINE with dgl

### PR DESCRIPTION
**Pull request by ShawXh**
## Description

Implement a network embedding algorithm, LINE (https://arxiv.org/pdf/1503.03578). The implementation has been tested on several ogbn datasets. The results are as following:

ogbl-arxiv
#params: 33023343(model) + 142888(mlp) = 33166231
Highest Train: 82.94 ± 0.11
Highest Valid: 71.76 ± 0.08
Final Train: 80.74 ± 1.30
Final Test: 70.47 ± 0.19

obgl-proteins
#params: 25853524(model) + 129648(mlp) = 25983172
Highest Train: 93.11 ± 0.04
Highest Valid: 70.50 ± 1.29
Final Train: 77.66 ± 10.27
Final Test: 62.07 ± 1.25

ogbl-products
#params: 477570049(model) + 136495(mlp) = 477706544
Highest Train: 98.01 ± 0.32
Highest Valid: 89.57 ± 0.09
Final Train: 94.96 ± 0.43
Final Test: 72.52 ± 0.29

The results are also given in README.md. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/ogb/line/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+- Official implementation: [here](https://github.com/tangjianpku/LINE)
+
+## How to load ogb data
+To load ogb dataset, you need to run the following command, which will output a network file, ogbn-products-net.txt:
+```
+python3 load_dataset.py --name ogbn-proteins
+```
+Or you can run the code directly with:
+```
+python3 deepwalk --ogbn_name xxx --load_from_ogbn
```
Comment (classicsong) [inline examples/pytorch/ogb/line/README.md:12]: why still deepwalk?

File: examples/pytorch/ogb/line/line.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        """ Initializing the trainer with the input arguments """
+        self.args = args
+        self.dataset = LineDataset(
+            net_file=args.data_file,
+            map_file=args.map_file,
+            batch_size=args.batch_size,
+            negative=args.negative,
+            gpus=args.gpus,
+            fast_neg=args.fast_neg,
+            ogbl_name=args.ogbl_name,
```
Comment (classicsong) [inline examples/pytorch/ogb/line/line.py:26]: can we simplify these 4 args into 1?

File: examples/pytorch/ogb/line/line.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+            assert self.args.gpus[0] >= 0
+            self.emb_model.all_to_device(self.args.gpus[0])
+        elif self.args.mix:
+            print("Mix CPU with %d GPU" % len(self.args.gpus))
+            if len(self.args.gpus) == 1:
+                assert self.args.gpus[0] >= 0, 'mix CPU with GPU should have avaliable GPU'
+                self.emb_model.set_device(self.args.gpus[0])
+        else:
+            print("Run in CPU process")
+            self.args.gpus = [torch.device('cpu')]
```
Comment (classicsong) [inline examples/pytorch/ogb/line/line.py:74]: better be self.device or just device.

File: examples/pytorch/ogb/line/load_dataset.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    else:
+        g = load_from_ogbn_with_name(name=name)
+
+    try:
+        w = g.edata['edge_weight']
+        weighted = True
+    except:
+        weighted = False
+
+    with open(name + "-net.txt", "w") as f:
```
Comment (classicsong) [inline examples/pytorch/ogb/line/load_dataset.py:41]: there is a dgl.save_graph dgl.load_graph API.

File: examples/pytorch/ogb/line/model.py [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+        grad_sum = grad_sum.cpu()
+    state_sum.index_add_(0, nodes, grad_sum) # cpu
+    std = state_sum[nodes].to(device)  # gpu
+    std_values = std.sqrt_().add_(1e-10).unsqueeze(1)
+    grad = (lr * grad / std_values) # gpu
+
+    return grad
+
+@thread_wrapped_func
+def async_update(num_threads, model, queue):
```
Comment (classicsong) [inline examples/pytorch/ogb/line/model.py:48]: Does async_update introduce speedup?

File: examples/pytorch/ogb/line/model.py [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+        """ For ogb leaderboard. """
+        u_embeddings = torch.Tensor(self.get_embedding())
+        embedding = torch.zeros_like(u_embeddings.cpu().data)
+        valid_nodes = torch.LongTensor(dataset.valid_nodes)
+        valid_embedding = u_embeddings.cpu().data.index_select(0, valid_nodes)
+        embedding.index_add_(0, valid_nodes, valid_embedding)
+
+        torch.save(embedding, file_name)
+
+    def save_embedding_txt(self, dataset, file_name):
```
Comment (classicsong) [inline examples/pytorch/ogb/line/model.py:489]: I think we can remove this.

File: examples/pytorch/ogb/line/README.md (hunk 1 above)
Comment (ShawXh) [reply examples/pytorch/ogb/line/README.md:12]: my bad...

File: examples/pytorch/ogb/line/load_dataset.py (hunk 4 above)
Comment (ShawXh) [reply examples/pytorch/ogb/line/load_dataset.py:41]: ok. I will remove this

File: examples/pytorch/ogb/line/model.py (hunk 6 above)
Comment (ShawXh) [reply examples/pytorch/ogb/line/model.py:489]: ok

File: examples/pytorch/ogb/line/model.py (hunk 5 above)
Comment (ShawXh) [reply examples/pytorch/ogb/line/model.py:48]: seems not ...

**Comment by yzh119** [comment]
@ShawXh why closing this PR?

**Comment by ShawXh** [comment]
> @ShawXh why closing this PR?

I have pushed something irrelevant to this PR while I don't know how to delete the irrelevant thing... I think maybe I should open another PR...

**Pull request closed without merging**